### PR TITLE
Fix for out of bound read in std::strtol while parsing HTTP requests

### PR DESCRIPTION
**Pull request by alexprabhat99**
Fixes: #1193 
This is an attempt to better a previous PR #1201.
@kiplingw This PR is ready for review. Thanks!

### CODE REVIEW

**Comment by kiplingw** [comment]
@Tachi107, I can't tell, but are these CI failures innocuous?

File: src/common/http.cc [hunk 1]
```diff
@@ -287,9 +288,18 @@ namespace Pistache::Http
...
-            if (*end != ' ')
+            const char* beg = codeToken.rawText();
+            const char* beg_trim = beg;
+            while (*beg_trim == '+' || *beg_trim == ' ')
+            {
+                if (*(beg_trim + 1))
+                    ++beg_trim;                    
+            }
+            std::string_view view(beg_trim, codeToken.size());
+            const char* end = view.data() + view.size();
```
Comment (Tachi107) [inline src/common/http.cc:299]: Why is a string_view being used here?

Comment (Tachi107) [reply src/common/http.cc:299]: Citing cppreference:

>     constexpr basic_string_view( const CharT* s, size_type count );
>
> After construction, [data()](https://en.cppreference.com/w/cpp/string/basic_string_view/data) is equal to s, and [size()](https://en.cppreference.com/w/cpp/string/basic_string_view/size) is equal to count.

So isn't your code exactly the same as:

```suggestion
            const char* end = beg_trim + codeToken.size();
```

File: src/common/http.cc [hunk 2]
```diff
@@ -455,10 +465,18 @@ namespace Pistache::Http
...
-                if (*end != '\r')
+                const char* beg = chunkSize.rawText();
+                const char* beg_trim = beg;
+                while (*beg_trim == '+' || *beg_trim == ' ')
+                {
+                    if (*(beg_trim + 1))
+                        ++beg_trim; 
+                }
+                std::string_view view = beg_trim;
+                const char* end = view.data() + view.size();
```
Comment (Tachi107) [inline src/common/http.cc:476]: I'm generally a fan of string_view, but here I personally find it unnecessary. I'd suggest using `strlen()` explicitly, like:

```suggestion
                const char* end = beg_trim + std::strlen(beg_trim);
```

**Review by Tachi107 — CHANGES_REQUESTED** [review]
This patch is supposed to be fixing an out of bounds read due to using `std::strtol` on a non-`\0`-terminated string, but unless I'm reading this wrong the bug is still present; instead of over-reading with `strtol`, you're over-reading with `std::from_chars`, since the `end` pointer is derived by `beg_trim + strlen(beg_trim)`, which will point to beyond the end of the string when it isn't null-terminated since `strlen` looks for the string terminator to determined the size.

If you have a char pointer and the pointed data isn't null-terminated *and* you don't even have a length/size variable, then you really have no way of determining its length.

**Comment by Tachi107** [comment]
Il giorno ven 26 apr 2024 alle 10:52:45 -07:00:00, Kip 
***@***.***> ha scritto:
> @Tachi107, I can't tell, but are these CI failures innocuous?

It seems an error in the runners themselves, so not related to this 
patch in particular. I've just rescheduled them, and they should 
hopefully pass.

-- 
OpenPGP key: 66DE F152 8299 0C21 99EF  A801 A8A1 28A8 AB1C EE49

**Comment by kiplingw** [comment]
Fixed with #1251.

**Pull request closed without merging**